### PR TITLE
feat: add theme support

### PR DESCRIPTION
**Pull request by T3kstiil3**
Add new resume template 👌

image

image

### CODE REVIEW

**Comment by CapsHil** [comment]
Hello Aurélien,
J'ai eu l'occasion d'échanger avec le service marketing sur le sujet, on c'était dit qu'avoir le header en noir avec le logo Znk blanc c'était sympa aussi !

**Comment by T3kstiil3** [comment]
Hello, pas de souci pour faire d'autres styles de cv. Des dispo sur slack / hangout pour en discuter ?

File: app/components/Footer.jsx [hunk 1]
```diff
@@ -1,15 +1,23 @@
...
 const buildPath = require('../../build-path');
 
+
 const { string, func } = PropTypes;
 
 export default class Footer extends Component {
   constructor(props) {
     super(props);
+    this.state = {
+      themesBarState: "",
```
Comment (Gillespie59) [inline app/components/Footer.jsx]: Why you do not use a Boolean value ?

File: app/components/Footer.jsx [hunk 2]
```diff
@@ -35,10 +43,31 @@ export default class Footer extends Component {
...
   render() {
     let path = '';
 
-    if (this.props.metadata) {
+    if (this.props.path) {
+      path = this.props.path;
+    }
+    else if (this.props.metadata) {
+      if(this.props.metadata.firstname){
+        path = buildPath(this.props.metadata.name + ' ' + this.props.metadata.firstname + '');
```
Comment (Gillespie59) [inline app/components/Footer.jsx]: Prefer string interpolation

File: app/components/Footer.jsx [hunk 3]
```diff
@@ -67,13 +97,37 @@ export default class Footer extends Component {
...
+          >Themes</button>
+        </span>
+        <div
+          className={`themes ${this.state.themesBarState}`}
+        >
+          <h3>Themes</h3>
+          <ul>
+            {this.themes.map((theme, index) => {
+              return (
+                <li key={index} className="theme-item" onClick={() => this.props.changeTheme(theme)}>
```
Comment (Gillespie59) [inline app/components/Footer.jsx]: A li is not clickable. Use a button , with a reset CSS, inside of the li.

File: app/components/Footer.jsx [hunk 4]
```diff
@@ -67,13 +97,37 @@ export default class Footer extends Component {
...
+                  <img src={require(`../static/img/theme-${theme.name}.png`)} alt="default" />
+                  <span>{theme.name} <i className="fa fa-info"></i></span>
+                </li>
+              )
+            })}
+          </ul>
+          <a className="close" title="close theme list" onClick={this.toggleThemesBarState} href="#">
```
Comment (Gillespie59) [inline app/components/Footer.jsx]: A link with a href=# is not a link.. use a button instead

File: server.js [hunk 5]
```diff
@@ -271,7 +271,11 @@ api.put('/documents/:uuid', bodyParser.json(), (req, res) => {
         sql = 'UPDATE resume SET content = $1, path = $3, version = $4, last_modified = $5, metadata = $6 where uuid = $2';
       }
 
-      var path = buildPath(req.body.metadata.name + '');
+      if(req.body.metadata.firstname){
+        var path = buildPath(req.body.metadata.name + ' ' + req.body.metadata.firstname + '');
```
Comment (Gillespie59) [inline server.js]: Sting interpolation

Comment (Gillespie59) [inline server.js]: Use let instead

File: server.js [hunk 6]
```diff
@@ -271,7 +271,11 @@ api.put('/documents/:uuid', bodyParser.json(), (req, res) => {
...
+      }else {
+        var path = buildPath(req.body.metadata.name);
```
Comment (Gillespie59) [inline server.js]: let

File: app/components/Footer.jsx [hunk 7]
```diff
@@ -35,10 +43,31 @@ export default class Footer extends Component {
...
+    if (this.props.path) {
+      path = this.props.path;
+    }
+    else if (this.props.metadata) {
+      if(this.props.metadata.firstname){
+        path = buildPath(`${this.props.metadata.name} ${this.props.metadata.firstname}`);
+      }else {
+        path = buildPath(this.props.metadata.name);
+      }
       path = buildPath(this.props.metadata.name);
```
Comment (hgwood) [inline app/components/Footer.jsx]: Isn't this reverting the effect of lines 67 and 69?

File: app/components/Footer.jsx [hunk 8]
```diff
@@ -67,13 +97,37 @@ export default class Footer extends Component {
...
+            className="showThemesBtn"
+            onClick={this.toggleIsThemesBarOpen}
+          >Themes</button>
+        </span>
+        <div className={`themes ${this.state.isThemesBarOpen ? 'open' : ''}`} >
+          <h3>Themes</h3>
+          <ul>
+            {this.themes.map((theme, index) => {
+              return (
+                <li key={index} className="theme-item">
```
Comment (hgwood) [inline app/components/Footer.jsx]: Using a the theme's name as the key would make more sense, because the order of the themes in the array is not significant, as far as I can see.

File: app/components/Footer.jsx [hunk 9]
```diff
@@ -35,10 +43,31 @@ export default class Footer extends Component {
...
+      this.setState({
+        isThemesBarOpen: false,
+      });
+    }
+    else {
+      this.setState({
+        isThemesBarOpen: true,
+      });
+    }
+  }
```
Comment (hgwood) [inline app/components/Footer.jsx]: Maybe this could work?

```js
this.setState({
  isThemeBarOpen: !this.state.isThemeBarOpen,
});
```

File: app/components/Footer.jsx [hunk 10]
```diff
@@ -35,10 +43,31 @@ export default class Footer extends Component {
...
-    if (this.props.metadata) {
+    if (this.props.path) {
+      path = this.props.path;
+    }
+    else if (this.props.metadata) {
+      if(this.props.metadata.firstname){
+        path = buildPath(`${this.props.metadata.name} ${this.props.metadata.firstname}`);
+      }else {
+        path = buildPath(this.props.metadata.name);
+      }
```
Comment (hgwood) [inline app/components/Footer.jsx]: This logic is getting complex and I think it would be worth it to extract it into its own function that returns the path.

File: app/components/Editor.jsx [hunk 11]
```diff
@@ -19,7 +20,7 @@ export default class Editor extends Component {
     this.state = {
       template: 'resume',
       pos: 0,
-      mode: EditorModes.PREVIEW
+      mode: localStorage.getItem('LastEditorModesSet') || EditorModes.PREVIEW
```
Comment (hgwood) [inline app/components/Editor.jsx]: How is this related to themes?

File: app/components/Preview.jsx [hunk 12]
```diff
@@ -120,6 +121,18 @@ export default class Preview extends Component {
     return chunks;
   }
 
+  getClassTheme() {
```
Comment (hgwood) [inline app/components/Preview.jsx]: Did you mean `getThemeClass`? `getThemeCssClass` would be even better I think.

File: app/components/PreviewFlag.jsx [hunk 13]
```diff
@@ -15,7 +16,7 @@ class PreviewChunk extends Component {
       }
     });
 
-    return !isEqual(this.props.chunk, nextProps.chunk) || this.props.key !== nextProps.key;
+    return !isEqual(this.props.chunk, nextProps.chunk) || this.props.id !== nextProps.id;
```
Comment (hgwood) [inline app/components/PreviewFlag.jsx]: How is this related to themes?

File: app/components/templates/Resume.jsx [hunk 14]
```diff
@@ -111,25 +121,49 @@ export default class Resume extends BaseTemplate {
...
-    );
+    return table;
   }
 
   render() {
     const data = this.getData();
 
     const expertises = this.createList(data, 'expertises');
 
+    if (this.props.data.theme && this.props.data.theme == "light") {
```
Comment (hgwood) [inline app/components/templates/Resume.jsx]: `this.props.data.theme === "light"` should be enough of a condition :)

File: app/components/templates/Resume.jsx [hunk 15]
```diff
@@ -111,25 +121,49 @@ export default class Resume extends BaseTemplate {
...
+      return (
+        <div>
+
+          <div className="topbarre"></div>
```
Comment (hgwood) [inline app/components/templates/Resume.jsx]: Did you mean `topbar`?

File: database.sql [hunk 16]
```diff
@@ -1 +1 @@
-create table resume (id serial primary key, content text, metadata text, uuid text, path text, version integer, last_modified timestamp);
\ No newline at end of file
+CREATE TABLE resume (id serial primary key, content text, metadata text, uuid text, path text, version integer, last_modified timestamp);
```
Comment (hgwood) [inline database.sql]: How is this related to themes?

File: server.js [hunk 17]
```diff
@@ -259,19 +259,25 @@ api.put('/documents/:uuid', bodyParser.json(), (req, res) => {
...
         sql = 'UPDATE resume SET content = $1, path = $3, version = $4, last_modified = $5, metadata = $6 where uuid = $2';
       }
 
-      var path = buildPath(req.body.metadata.name + '');
+      let path = '';
+
+      if(req.body.metadata.firstname){
+        path = buildPath(`${req.body.metadata.name} ${req.body.metadata.firstname}`);
+      }else {
+        path = buildPath(req.body.metadata.name);
```
Comment (hgwood) [inline server.js]: Before the change, there was a conversion to string here (using `+ ''`). I don't know why it was there, but wouldn't it be best to keep it?

File: server.js [hunk 18]
```diff
@@ -259,19 +259,25 @@ api.put('/documents/:uuid', bodyParser.json(), (req, res) => {
...
+      }
```
Comment (hgwood) [inline server.js]: A ternary expression would allow for `path` to be a `const`.

File: server.js [hunk 19]
```diff
@@ -20,12 +20,12 @@ const GoogleStrategy = require('passport-google-oauth20').Strategy;
...
 const staticPath = path.join(__dirname, '/build');
 
-var databaseUrl = process.env.DATABASE_URL || 'postgres://localhost/resume';
-var googleId = process.env.GOOGLE_ID || require('./conf-google').id;
-var googleSecret = process.env.GOOGLE_SECRET || require('./conf-google').secret;
-var googleCallback = process.env.GOOGLE_CALLBACK || require('./conf-google').callback;
+const databaseUrl = process.env.DATABASE_URL || 'postgres://localhost/resume';
+const googleId = process.env.GOOGLE_ID || require('./conf-google').id;
+const googleSecret = process.env.GOOGLE_SECRET || require('./conf-google').secret;
+const googleCallback = process.env.GOOGLE_CALLBACK || require('./conf-google').callback;
```
Comment (hgwood) [inline server.js]: 👍

File: app/components/PreviewFlag.jsx (hunk 13 above)
Comment (T3kstiil3) [reply app/components/PreviewFlag.jsx]: Migration react version :/

**Comment by hgwood** [comment]
I've merged #26, could you revert the modifications on `package.json` on this PR?

File: app/components/templates/Resume.jsx (hunk 14 above)
Comment (T3kstiil3) [reply app/components/templates/Resume.jsx]: Old resume does not have metadata key theme

Comment (hgwood) [reply app/components/templates/Resume.jsx]: Yes but then it will be `undefined` so no problem with checking it. :)
